### PR TITLE
[12.0][MIG] sale_global_discount : Migration to 12.0

### PR DESCRIPTION
**Pull request by manuelcalerosolis**
This is the migration of module : sale_global_discount to 12.0

### CODE REVIEW

**Comment by HaraldPanten** [comment]
Hi @manuelcalerosolis could you check reds?

**Comment by manuelcalerosolis** [comment]
@HaraldPanten the travis error are like this

************* Module sales_team_security
sales_team_security/readme/CONTRIBUTORS.rst:1: [W7908(missing-newline-extrafiles), ]  Missing newline
************* Module sale_exception.models.sale_order_line
sale_exception/models/sale_order_line.py:1: [C8202(unnecessary-utf8-coding-comment), ] UTF-8 coding is not necessary
************* Module portal_sale_personal_data_only.__manifest__

They are errors of other modules, so I do not understand anything. Can you help me know what's going on?

**Comment by HaraldPanten** [comment]
Hi @manuelcalerosolis I don't really have any idea. @ValentinVinagre could you have a look? Thx

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "restructuredtext.confPath": ""
+}
```
Comment (gurneyalex) [inline .vscode/settings.json]: can you remove this file?

File: sale_global_discount/models/sale_order.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            })
+        return res
+
+    @api.onchange('partner_id')
+    def onchange_partner_id(self):
+        res = super().onchange_partner_id()
+        if self.partner_id.customer_global_discount_ids:
+            self.global_discount_ids = (
+                self.partner_id.customer_global_discount_ids or
+                self.parnter_id.commercial_partner_id
```
Comment (gurneyalex) [inline sale_global_discount/models/sale_order.py]:
```suggestion
                self.partner_id.commercial_partner_id
```

Can you check if this bug is also in the original version?

File: sale_global_discount/models/sale_order.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+    @api.model
+    def get_discounted_global(self, price=0, discounts=None):
+        """Compute discounts successively"""
+        discounts = discounts or []
+        if not discounts:
+            return price
+        discount = discounts.pop(0)
+        price *= 1 - (discount / 100)
+        return self.get_discounted_global(price, discounts)
```
Comment (gurneyalex) [inline sale_global_discount/models/sale_order.py]: using a recursive method here is  inefficient, and it also forces you to make copies of dicounts in other places. 

```suggestion
    def get_discounted_global(self, price=0, discounts=None):
        """Compute discounts successively
        discounts is a list of discount rates, or None
        """
        if discounts is None:
            discounts = []
        for discount in discounts:
            price *= 1 - (discount / 100)
        return price
```

File: sale_global_discount/models/sale_order.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    def _amount_all(self):
+        res = super()._amount_all()
+        for order in self:
+            amount_untaxed_before_global_discounts = order.amount_untaxed
+            amount_total_before_global_discounts = order.amount_total
+            discounts = order.global_discount_ids.mapped('discount')
+            amount_discounted_untaxed = amount_discounted_tax = 0
+            for line in order.order_line:
+                discounted_subtotal = self.get_discounted_global(
+                    line.price_subtotal, discounts.copy())
```
Comment (gurneyalex) [inline sale_global_discount/models/sale_order.py]:
```suggestion
                    line.price_subtotal, discounts)
```

No copy required with the implementation suggested above

File: sale_global_discount/models/sale_order.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        """We can apply discounts directly by tax groups"""
+        tax_groups = self._get_tax_by_group()
+        discounts = self.global_discount_ids.mapped('discount')
+        if not discounts:
+            return tax_groups
+        round_curr = self.currency_id.round
+        res = []
+        for tax in tax_groups:
+            tax_amount = round_curr(
+                self.get_discounted_global(tax[1], discounts.copy()))
```
Comment (gurneyalex) [inline sale_global_discount/models/sale_order.py]:
```suggestion
                self.get_discounted_global(tax[1], discounts))
```

File: sale_global_discount/models/sale_order.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            tax_base = round_curr(
+                self.get_discounted_global(tax[2], discounts.copy()))
```
Comment (gurneyalex) [inline sale_global_discount/models/sale_order.py]:
```suggestion
                self.get_discounted_global(tax[2], discounts))
```

**Review by gurneyalex — CHANGES_REQUESTED** [review]

File: sale_global_discount/models/sale_order.py (hunk 2 above)
Comment (manuelcalerosolis) [reply sale_global_discount/models/sale_order.py]: Yes, I have suggested the module author to correct it.

**Comment by rousseldenis** [comment]
@manuelcalerosolis could you edit PR comment and say it depends on https://github.com/OCA/account-invoicing/pull/719 ?

**Comment by rousseldenis** [comment]
@manuelcalerosolis For travis errors, see lint line and flake8 specific errors

**Comment by pedrobaeza** [comment]
@manuelcalerosolis are you going to fix this PR?

**Comment by manuelcalerosolis** [comment]
Yes of course, as soon as possible.

**Comment by pedrobaeza** [comment]
Check the fixes made in #1181 (and its counterpart for account), that are needed to be correct with several taxes involved.

**Comment by pedrobaeza** [comment]
Please don't perform merge commits in the PR, as they screw the commit history for later migrations.

File: sale_global_discount/models/sale_order.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        discounts = self.global_discount_ids.mapped('discount')
+        if not discounts:
+            return tax_groups
+        round_curr = self.currency_id.round
+        res = []
+        for tax in tax_groups:
+            tax_amount = round_curr(
+                self.get_discounted_global(tax[1], discounts))
+            tax_base = round_curr(
+                self.get_discounted_global(tax[2], discounts)
```
Comment (acysos) [inline sale_global_discount/models/sale_order.py]: Fail end function:
tax_base = round_curr(
                self.get_discounted_global(tax[2], discounts))

**Review by acysos — CHANGES_REQUESTED** [review]

File: sale_global_discount/models/sale_order.py (hunk 7 above)
Comment (manuelcalerosolis) [reply sale_global_discount/models/sale_order.py]: @acysos Fixed!

Thanks.

**Comment by pedrobaeza** [comment]
Superseded by #1197

**Pull request closed without merging**